File: android/AppPlatformReactNativeApp/appplatformble/BleTypes.kt
package com.AppPlatformReactNativeApp.trakplusble
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
enum class ConnectionState(val wireValue: String) {
  DISCONNECTED("DISCONNECTED"),
  CONNECTING("CONNECTING"),
  CONNECTED("CONNECTED"),
  DISCOVERING("DISCOVERING"),
  SUBSCRIBING("SUBSCRIBING"),
  READY("READY");
}
enum class DisconnectReason(val wireValue: String) {
  NORMAL("normal"),
  BONDING_LOST("bondingLost"),
  ENCRYPTION_FAILED("encryptionFailed"),
  CONNECTION_FAILED("connectionFailed"),
  TIMEOUT("timeout"),
  UNKNOWN("unknown"),
  DEVICE_SLEEP("deviceSleep"),
  SERVICE_DISCOVERY_FAILED("serviceDiscoveryFailed"),
  SERVICE_NOT_FOUND("serviceNotFound"),
  CHARACTERISTIC_DISCOVERY_FAILED("characteristicDiscoveryFailed"),
  CHARACTERISTIC_NOT_FOUND("characteristicNotFound"),
  SUBSCRIPTION_FAILED("subscriptionFailed"),
  SUBSCRIPTION_LOST("subscriptionLost"),
  DISCOVERY_TIMEOUT("discoveryTimeout"),
  SUBSCRIPTION_TIMEOUT("subscriptionTimeout");
}
enum class BluetoothAdapterState(val wireValue: String) {
  UNKNOWN("unknown"),
  RESETTING("resetting"),
  UNSUPPORTED("unsupported"),
  UNAUTHORIZED("unauthorized"),
  POWERED_OFF("poweredOff"),
  POWERED_ON("poweredOn");
}
enum class OperationType(val wireValue: String) {
  CONNECT("connect"),
  WRITE("write");
}
enum class OperationRejectReason(val wireValue: String) {
  NOT_INITIALIZED("notInitialized"),
  UNAUTHORIZED("unauthorized"),
  UNSUPPORTED("unsupported"),
  BUSY("busy"),
  NOT_READY("notReady"),
  INVALID_PAYLOAD("invalidPayload"),
  QUEUE_FULL("queueFull");
}
data class OperationRejectedEvent(
  val operation: OperationType,
  val reason: OperationRejectReason,
  val detail: String? = null,
  val deviceId: String? = null,
  val timestampMs: Long = System.currentTimeMillis(),
)
data class ConnectionStateEvent(
  val state: ConnectionState,
  val deviceId: String,
  val reason: DisconnectReason? = null,
  val timestampMs: Long = System.currentTimeMillis(),
  val foregroundServiceActive: Boolean = false,
) {
  fun toWritableMap(): WritableMap {
    val map = Arguments.createMap()
    map.putString("state", state.wireValue)
    map.putString("deviceId", deviceId)
    if (reason != null) {
      map.putString("reason", reason.wireValue)
    }
    map.putDouble("timestamp", timestampMs.toDouble())
    map.putBoolean("foregroundServiceActive", foregroundServiceActive)
    return map
  }
}
data class DataReceivedEvent(
  val dataBase64: String,
  val deviceId: String,
  val timestampMs: Long = System.currentTimeMillis(),
) {
  fun toWritableMap(): WritableMap {
    val map = Arguments.createMap()
    map.putString("data", dataBase64)
    map.putString("deviceId", deviceId)
    map.putDouble("timestamp", timestampMs.toDouble())
    return map
  }
}
data class DiscoveredDevice(
  val id: String,
  val name: String?,
  val rssi: Int,
  val isConnectable: Boolean,
) {
  fun toWritableMap(): WritableMap {
    val map = Arguments.createMap()
    map.putString("id", id)
    map.putString("name", name)
    map.putInt("rssi", rssi)
    map.putBoolean("isConnectable", isConnectable)
    return map
  }
}
object BleConfig {
  const val CONNECTION_TIMEOUT_MS = 30_000L
  const val DISCOVERY_TIMEOUT_MS = 10_000L
  const val SUBSCRIPTION_TIMEOUT_MS = 5_000L
  const val WRITE_TIMEOUT_MS = 5_000L
  const val BOND_TIMEOUT_MS = 20_000L
  const val MAX_RECONNECTION_ATTEMPTS = 5
  const val DORMANT_RECONNECTION_INTERVAL_MS = 60_000L
  const val RECONNECT_BASE_DELAY_MS = 1_000L
  const val MAX_WRITE_QUEUE_SIZE = 128
  const val STABLE_CONNECTION_THRESHOLD_MS = 55_000L
  const val REQUESTED_MTU = 247
  const val SERVICE_UUID = "4fafc201-1fb5-459e-8fcc-c5c9c331914b"
  const val MAIN_CHARACTERISTIC_UUID = "beb5483e-36e1-4688-b7f5-ea07361b26a8"
  const val CLIENT_CHARACTERISTIC_CONFIG_UUID = "00002902-0000-1000-8000-00805f9b34fb"
}

File: ios/AppPlatform/AppPlatformBLE/BLETypes.swift
import Foundation
import CoreBluetooth
/
/
enum AppDeviceUUIDs {
    /
    static let service = CBUUID(string: "4fafc201-1fb5-459e-8fcc-c5c9c331914b")
    /
    static let mainCharacteristic = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26a8")
    /
    static let otaCharacteristic = CBUUID(string: "beb5483e-36e1-4688-b7f5-ea07361b26a9")
}
/
/
/
/
/
/
/
/
/
enum ConnectionState: String, Codable {
    /
    case disconnected = "DISCONNECTED"
    /
    /
    case connecting = "CONNECTING"
    /
    /
    case connected = "CONNECTED"
    /
    /
    case discovering = "DISCOVERING"
    /
    /
    case subscribing = "SUBSCRIBING"
    /
    /
    case ready = "READY"
}
/
/
/
/
/
/
/
/
/
/
/
/
/
/
enum DisconnectReason: String, Codable {
    /
    case normal = "normal"
    /
    /
    case bondingLost = "bondingLost"
    /
    /
    case encryptionFailed = "encryptionFailed"
    /
    /
    case connectionFailed = "connectionFailed"
    /
    case timeout = "timeout"
    /
    case unknown = "unknown"
    /
    case deviceSleep = "deviceSleep"
    /
    case serviceDiscoveryFailed = "serviceDiscoveryFailed"
    /
    /
    case serviceNotFound = "serviceNotFound"
    /
    case characteristicDiscoveryFailed = "characteristicDiscoveryFailed"
    /
    /
    case characteristicNotFound = "characteristicNotFound"
    /
    /
    case subscriptionFailed = "subscriptionFailed"
    /
    /
    case subscriptionLost = "subscriptionLost"
    /
    case discoveryTimeout = "discoveryTimeout"
    /
    case subscriptionTimeout = "subscriptionTimeout"
}
/
enum BluetoothAdapterState: String, Codable {
    case unknown = "unknown"
    case resetting = "resetting"
    case unsupported = "unsupported"
    case unauthorized = "unauthorized"
    case poweredOff = "poweredOff"
    case poweredOn = "poweredOn"
    init(from cbState: CBManagerState) {
        switch cbState {
        case .unknown: self = .unknown
        case .resetting: self = .resetting
        case .unsupported: self = .unsupported
        case .unauthorized: self = .unauthorized
        case .poweredOff: self = .poweredOff
        case .poweredOn: self = .poweredOn
        @unknown default: self = .unknown
        }
    }
}
/
struct DiscoveredDevice: Codable {
    let id: String
    let name: String?
    let rssi: Int
    let isConnectable: Bool
}
/
struct ConnectionStateEvent: Codable {
    let state: ConnectionState
    let deviceId: String
    let reason: DisconnectReason?
    let timestamp: TimeInterval
    init(state: ConnectionState, deviceId: String, reason: DisconnectReason? = nil) {
        self.state = state
        self.deviceId = deviceId
        self.reason = reason
        self.timestamp = Date().timeIntervalSince1970 * 1000 
    }
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "state": state.rawValue,
            "deviceId": deviceId,
            "timestamp": timestamp
        ]
        if let reason = reason {
            dict["reason"] = reason.rawValue
        }
        return dict
    }
}
/
struct DataReceivedEvent: Codable {
    let data: String  
    let deviceId: String
    let timestamp: TimeInterval
    init(data: Data, deviceId: String) {
        self.data = data.base64EncodedString()
        self.deviceId = deviceId
        self.timestamp = Date().timeIntervalSince1970 * 1000
    }
    func toDictionary() -> [String: Any] {
        return [
            "data": data,
            "deviceId": deviceId,
            "timestamp": timestamp
        ]
    }
}
/
/
protocol AppDeviceBLECoreDelegate: AnyObject {
    /
    func connectionStateChanged(_ event: ConnectionStateEvent)
    /
    func dataReceived(_ event: DataReceivedEvent)
    /
    func bondingLost(deviceId: String)
    /
    func bluetoothStateChanged(_ state: BluetoothAdapterState)
    /
    func deviceDiscovered(_ device: DiscoveredDevice)
    /
    /
    func operationRejected(_ event: OperationRejectedEvent)
}
/
/
enum OperationType: String, Codable {
    case connect = "connect"
    case write = "write"
}
/
/
enum OperationRejectReason: String, Codable {
    case notInitialized = "notInitialized"
    case unauthorized = "unauthorized"
    case unsupported = "unsupported"
    case busy = "busy"
    case notReady = "notReady"
    case invalidPayload = "invalidPayload"
    case queueFull = "queueFull"
}
/
/
/
/
/
struct OperationRejectedEvent {
    let operation: OperationType
    let reason: OperationRejectReason
    let detail: String?
    let deviceId: String?
    let timestampMs: TimeInterval
    init(
        operation: OperationType,
        reason: OperationRejectReason,
        detail: String? = nil,
        deviceId: String? = nil
    ) {
        self.operation = operation
        self.reason = reason
        self.detail = detail
        self.deviceId = deviceId
        self.timestampMs = Date().timeIntervalSince1970 * 1000
    }
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "operation": operation.rawValue,
            "reason": reason.rawValue,
            "timestamp": timestampMs
        ]
        if let detail = detail {
            dict["detail"] = detail
        }
        if let deviceId = deviceId {
            dict["deviceId"] = deviceId
        }
        return dict
    }
}
/
enum BLEConfig {
    /
    static let requestedMTU: Int = 512
    /
    static let connectionTimeoutSeconds: TimeInterval = 30
    /
    static let discoveryTimeoutSeconds: TimeInterval = 10
    /
    static let subscriptionTimeoutSeconds: TimeInterval = 5
    /
    static let restorationIdentifier = "AppDeviceBLE_Restore"
    /
    static let maxReconnectionAttempts = 5
    /
    static let dormantReconnectionIntervalSeconds: TimeInterval = 60
}
